src/librustc_mir/const_eval.rs

```diff
@@ -489,7 +489,12 @@ pub fn const_field<'a, 'tcx>(
     trace!("const_field: {:?}, {:?}", field, value);
     let ecx = mk_eval_cx(tcx, DUMMY_SP, param_env);
     // get the operand again
-    let op = ecx.eval_const_to_op(value, None).unwrap();
+    let mut op = ecx.eval_const_to_op(value, None).unwrap();
+    // adjust the alignment of `op` to the one of the allocation, since it may be a field of a
+    // packed struct and thus end up causing an alignment error if we read from it.
+    if let ConstValue::ByRef(_, alloc) = value.val {
+        op.force_alignment(alloc.align);
```

Comment:
I don't think this is enough in the general case. Can't we have a packed struct with fields `(u8, u16)` wrapped in an `aligned(2)` struct? Then `alloc.align` will be 2 but the alignment of the second field will still be wrong. The offset is 1, so when it gets access with the `u16` alignment we get an error.

Reply:
I added a test, but I wasn't able to trigger any problems

Reply:
I am not sure how to even trigger this code, but looking at the alignment of the allocation without taking into account the offset definitely makes no sense. So this code cannot remain the way it is, unless you convince me that I am wrong about this. ;)

Reply:
Actually I am wondering when this would even be necessary... for this to cause a problem, the original `op` must already be wrong, right?

Reply:
The problem is not `eval_const_to_op` producing bad `OpTy`s, it's `const_field` producing bad `ty::Const`s, which are later fed to `const_field` *again* and then cause `eval_const_to_op` to produce a bad `OpTy` because the input was already bad.

I mean this is trivially fixable by giving `ByRef` an `Align` field, but it's unnecessary for everything but `const_field`. And in `const_field` this only ever has an effect when doing pattern match checking. I pushed a fix that just ignores alignment in `const_field`. I mean that is what happens for `Immediate` aggregates, too.

Reply:
> I mean that is what happens for Immediate aggregates, too.

... which may be why I failed to write a repro test :D

Reply:
> it's const_field producing bad ty::Consts, which are later fed to const_field again and then cause eval_const_to_op to produce a bad OpTy because the input was already bad.

Okay. But shouldn't this happen on the *return* value of `const_field` then if that's already wrong?

Reply:
that's what we used to have, we generated a new `Allocation` with the appropriate alignment. We could extend `ConstValue::ByRef` to also have an `Align` field so we can properly propagate this information. It would be unused by anything other than `const_field` though. It won't regress the size of `ConstValue` I believe, so we could just do it.